마크다운 문법 v2 구현

RBJH opened the pull request:
- 각 문법에 맞는 HTML 태그 적용을 통해 WYSIWYG 구현
  - 기존 버전은 input 태그에 각 HTML 태그의 스타일을 적용
- 각 태그에 input 속성을 적용하기 위해 contentEditable 사용
  - [Content Editable](https://developer.mozilla.org/ko/docs/Web/Guide/HTML/Content_Editable)
- placehoder 적용을 위해 가상 클래스(:empty) 등 활용
  - [Supported CSS](https://www.styled-components.com/docs/api)

RBJH committed:
Fix: 코드 리뷰 사항 반영

변수의 정확성을 위해 enum 변수명 구체화(TYPE -> BUTTON_TYPE)
타입 체크 시 코드 가독성을 위해 boolean 타입 변수 활용(isTerminal)
불필요한 코드 제거

RBJH committed:
Chore: 테스트 위한 컴포넌트 렌더링

테스트를 위해 EditorPage에 구현중인 EditorInput 렌더링

RBJH committed:
Chore: ToolBarButton의 key prop 추가

Warning을 해결하기 위해 props에 key={index} 추가

RBJH committed:
Refactor: EditorInput 중복 코드 제거

enum 변수의 중복 코드를 제거하기 위해 MARKDOWN_TYPE 삭제
삭제한 부분의 의존성을 MARKDOWN_ATTRIBUTE의 key를 TYPE으로 사용해 해결

RBJH committed:
Feat: EditorInput 컴포넌트 태그 구현

스타일 통일성을 위해 styled-componentsDml HTML 태그를 활용

RBJH committed:
Feat: 마크다운 문법 li 구현

리스트를 구현하기 위해 styled.li 추가

RBJH committed:
Style: styled-components 코드 스타일 수정

복잡한 로직을 간단히 하기 위해 ampersands(&) 및 nesting 스타일 적용

RBJH committed:
Fix: 마크다운 리스트 문법 추가

ordered 리스트를 지원하기 위해 기존 li 타입을 ul로 수정
각 ol, ul을 렌더링 하기 위해 조건부 렌더링 활용

RBJH committed:
Feat: 마크다운 인용 문법(> ) 구현

의미에 적합한 태그 사용을 위해 blockquote 사용
마크다운 인용 디자인을 위해 border-left 및 padding-left 속성 조정

DrizzlingCattus commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -1,106 +1,201 @@
 
-import MARKDOWN_TYPE from "./MARKDOWN_TYPE";
+const MARKDOWN_TYPE = {
+  "": styled.div`
```
여기 이름없는건 잘 작동하나요? 조금 위태롭게 보이네요

RBJH replied:
이후 커밋에서 수정했습니다.:+1:

DrizzlingCattus commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -1,106 +1,201 @@
-  diplay: "block",
-  fontWeight: "bold",
-};
+  h1: styled.h1`
```
타입이 대문자이면 h1같은 속성 이름도 대문자가 되는게 좋다고 생각해요

RBJH replied:
이후 커밋에서 제거된 enum 변수입니다.

DrizzlingCattus commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -1,106 +1,201 @@
+
+    "1.": { ...state, type: "ol", placeholder: "list" },
+
+    ">": { ...state, type: "blockquote", placeholder: "quote" },
```
여기 state들도 다 spread하면 비효율적이예요. 변환되는 시점에 이런 작업을 하는게 더 좋겠네요!

ImuruKevol replied:
제가 state를 cellState로 고친 것처럼 구체화시키면 좋겠습니다.

RBJH replied:
반영했습니다!:+1:

DrizzlingCattus commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -1,106 +1,201 @@
+
+    if (key === " ") {
+      if (!state.type) {
+        setState(STATE[state.value]);
```
STATE라는 이름도 너무 모호해요. 다른 이름을 같이 생각해보죠

RBJH replied:
이후 커밋에서 제거된 변수입니다.

DrizzlingCattus commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -1,106 +1,201 @@
+        contentEditable={true}
+        onInput={onInput}
+        onKeyPress={onKeyPress}
+      />
```
위쪽 부분들은 마크다운 컴포넌트가 생기면 교체 및 리펙토링할 수 있겠군요

RBJH replied:
넵!

DrizzlingCattus commented on client/src/components/editor/toolbar/ToolBar.jsx:
```diff
@@ -1,23 +1,21 @@
 const ToolBar = () => {
-  const types = ["new", "save", "load", "code", "share", "terminal"];
+  const getToolBarButtons = () => {
+    return Object.keys(BUTTON_TYPE).map((buttonType, index) => {
```
Object.keys 부분도 따로 빼서 변수 이름을 붙혀서 사용하는게 가독성이 훨씬 좋아지겠네요

dimsssss reviewed (commented):
고생하셨습니다

RBJH committed:
Refactor: EditorInput 컴포넌트 리팩토링

기존 중복 코드를 제거하기 위해 styled-components의 as 속성 사용
as 속성 사용으로 불필요한 enum 타입 변수 제거
HTML 태그 blockquote를 커스터마이징하기 위해 props로 스타일 적용
코드 리뷰 사항 반영하기 위해 isQuote의 boolean 타입 변수를 사용해 타입 체크

RBJH committed:
Refactor: EditorInput 컴포넌트 리팩토링

렌더링 될 때마다 변수 변경 반복을 막기 위해 enum 변수 선언 위치 변경
변수 변경 필요한 이벤트 핸들러에 전개 구문(...)을 활용해 필요 시에만 변경
HTML 태그 blockquote 부분 스타일 조정(0.4rem, 0.8rem -> 0.25rem, 0.5rem)

RBJH committed:
Fix: EditorInput 컴포넌트 props 타입 디자인 이슈 수정

기존 삼항연산자로 blockquote 외 모든 HTML 태그 요소를 div의 default
값으로 할당해 타 요소에 잘못 적용되는 부분을 수정하기 위해 if문 사용
